linked.js

```diff
@@ -0,0 +1,161 @@
+class Node {
+    constructor(value = null, next = null) {
+        this.value = value;
+        this.next = next;
+    }
+}
+
+class linkedList {
+    constructor() {
+        // this.head = null;
+        // this.tail = null;
+        // this.next = null;
+    }
+    appendValue(value) {
+        if (!this.tail) {
+            this.tail = new Node(value);
+            this.head = this.tail;
+        }
+        else {
+            let oldTail = this.tail;
+            this.tail = new Node(value);
+            oldTail.next = this.tail;
+            // this.tail.prev = oldTail;
+        }
+    }
+    prependValue(value) {
+        if (!this.head) {
+            this.head = new Node(value);
+            this.tail = this.head;
+        }
+        else {
+            let oldhead = this.head;
+            this.head = new Node(value);
+            this.head.next = oldhead;
+            // oldhead.prev = this.head;
+        }
+    }
+    getSize() {
+        let currentNode = this.head;
+        let size = 0;
+        while (currentNode) {
+            currentNode = currentNode.next
+            size++
+        }
+        return size;
+    }
+    getHead() {
+        return this.head;
+    }
+    getTail() {
+        return this.tail
+    }
+    at(index) {
+        let currentNode = this.head;
+        let i = 0;
+        while (currentNode) {
+            if (i === index) {
+                return currentNode
+            }
+            currentNode = currentNode.next
+            i++
+        }
+    }
+    pop() {
+        if (this.getSize() <= 1) {
+            this.head = null;
+            this.tail = null;
+            return;
+        }
+        let size = this.getSize();
+        let newTail = this.at(size - 2);
+        this.tail = newTail;
+        this.tail.next = null;
+    }
+    shift() {
+        if (this.getSize() <= 1) {
+            this.head = null;
+            this.tail = null;
+            return;
+        }
+        let newHead = this.at(1);
+        this.head = newHead;
+    }
+    containsValue(obj, value) {
+        for (var key in obj) {
+            if (typeof obj[key] === 'object') {
+                return this.containsValue(obj[key], value);
+            }
+
+            if (obj[key] === value) {
+                return true;
+            }
+        }
+        return false;
+    }
+    find(value) {
+        let i = 0;
+        let currentNode = this.head;
+        while (currentNode) {
+            if (currentNode.value === value) {
+                return i;
+            }
+            currentNode = currentNode.next;
+            i++;
+        }
+    }
+    toString() {
+        let str = '';
+        let currentNode = this.head;
+        while (currentNode) {
+            str += `(${currentNode.value})` + " " + '-> ';
+            currentNode = currentNode.next;
+        }
+        return str;
+    }
+    insertAt(value, index) {
+        if (index === 0) {
+            this.prependValue(value);
+        }
```

Comment:
In the spirit of single responsibilty, you wouldn't want to duplicate node creation in `appendValue`, `prependValue`, `insertAt`. If you're up for the challenge, try having an api such as:
```js
appendValue(value) {
  insertAt(this.size) ;
}

prependValue(value) {
  insertAt(value, 0);
}

insertAt(value, 0) {
// handles both append, prepend and insert
}
```